util/nrve-niche-payment-handler.py

```diff
@@ -179,51 +148,94 @@ def do_sc_notify(self, event):
                 sql = ("update `NicheAuctionInvoicePayment`\n"
                        "set transactionId = %s\n"
                        ", transactionDate = from_unixtime(%s)\n"
                        "where fromNeoAddress = %s\n"
                        "and nrveAmount = %s\n"
                        "and paymentStatus = 0\n"
                        "and transactionId is null;")
-                args = (tx_hash, timestamp, from_address, nrve_amount)
+                args = (tx_hash, block.Timestamp, from_address, nrve_amount)
 
                 # Create a new record
                 cursor.execute(sql, args)
 
                 if cursor.rowcount != 1:
-                    self.logger.error('Failed updating payment. Should not be possible since it was already locked for update: %s', event)
+                    subject = 'Failed updating payment. Should not be possible since it was already locked for update'
+                    self.logger.error(subject + ': %s', event)
+                    self.send_email(
+                        subject,
+                        self.format_error_message({"Transaction Id": tx_hash}, {"From Address": from_address}, {"NRVE Amount": nrve_amount})
+                    )
                     return
 
-                # send the NRVE to the payment storage address
-                self.transfer_payment(self.niche_payment_storage_address, nrve_amount)
-
                 self.send_email("Successful Niche Payment", log)
 
-            # connection is not autocommit by default. So you must commit to save
-            # your changes.
+            # connection is not autocommit by default. So you must commit to save your changes.
             connection.commit()
+
         except MySQLError as e:
             error_message = 'ERROR: event %s: {!r}, errno is {}'.format(event, e, e.args[0])
             self.logger.error(error_message)
             self.send_email('Niche Payment Error', error_message)
+
         finally:
             connection.close()
 
+    def validate_transaction(self, connection, event, from_address, nrve_amount, tx_hash):
+
+        try:
+            with connection.cursor() as cursor:
+                sql = (
+                    "select oid from `NicheAuctionInvoicePayment` where transactionId = %s;"
+                )
+                params = tx_hash
+                cursor.execute(sql, params)
+
+                if cursor.rowcount == 0:
+                    # Send refund email
+                    subject = 'Failed identifying payment. Returning to sender'
+                    self.logger.error(subject + ': %s', event)
+                    self.send_email(
+                        subject,
+                        self.format_error_message({"Transaction Id": tx_hash}, {"From Address": from_address}, {"NRVE Amount": nrve_amount})
+                    )
+                    return
+
+                elif cursor.rowcount == 1:
```

Comment:
Let's hope we don't get 2 or more results 😂 😉

Reply:
Good point though, right now we just return. Perhaps we should at least log out the impossible.

Reply:
Yep, agreed.